Index: client_cache.c
===================================================================
--- client_cache.c	(revision 268)
+++ client_cache.c	(working copy)
@@ -59,6 +59,13 @@
 
 //static uint64_t gpu_comp_time = 0;
 
+extern int lock_gpu_comp_task(int gtask_num);
+extern void unlock_gpu_comp_task(int gtask_index);
+
+
+extern int lock_gpu_init_task(int gtask_num);
+extern void unlock_gpu_init_task(int gtask_index);
+
 int all_gpu_comp_ratio = 500;
 static uint64_t gpu_comp_cnt = 0;
 uint64_t gpu_comp_init_time = 0;
@@ -1792,6 +1799,7 @@
 		skyfs_u32_t count, char * databuf, int comp_type, int64_t * changed_space)
 {
 	int rc = 0;
+	char * d_temp = NULL;
 	uint64_t total_rc = 0;
 	int found_buffer = 0;
         skyfs_C_gcompbuf_unit_t * tmp;
@@ -1883,6 +1891,8 @@
 
 				struct timeval t3, t4;
 				uint64_t Gpu_comp_time = 0;
+				int Gpu_task_index = 0;
+				int Gpu_task_num = ((int)random())%64;
 
 				
 				char *  cpu_comp_buf_ptr[SKYFS_GPU_COMPRESS_BUF_SIZE /SKYFS_OBJECT_NODE_SIZE];
@@ -1890,14 +1900,24 @@
 
 				// TODO: only changed here , just for test , will be changed back by mayl
 				//tmp->gpu_comp_ratio = 1000;
-				if(((tmp->comp_op_count % 20)==0)&& tmp->comp_op_count >= 20){
+				if(((tmp->comp_op_count % 5)==0)&& (tmp->comp_op_count >= 5)){
 					adjust_comp_ratio(tmp);
 					
 				}
 
 			
+				// mayl: TODO do sharpening 
+				/*
 
-				if(random()%1000> tmp->gpu_comp_ratio){
+				if(tmp->gpu_comp_ratio <= 250){
+					tmp->gpu_comp_ratio = tmp->gpu_comp_ratio/100;
+				}else if(tmp->gpu_comp_ratio >=800){
+					//tmp->gpu_comp_ratio = (998+tmp->gpu_comp_ratio/500);
+				}
+				if(tmp->gpu_comp_ratio <=0){
+					tmp->gpu_comp_ratio = 1;
+				}*/
+				if(((int)(random()%1000))> tmp->gpu_comp_ratio){
 
 
 					
@@ -1931,6 +1951,8 @@
 					
 					cpu_compress_time += comp_time;
 					cpu_compress_bytes += SKYFS_GPU_COMPRESS_BUF_SIZE;
+					SKYFS_ERROR("CPu compress %lu us,  %lu bytes\n",
+							cpu_compress_time, cpu_compress_bytes);
 
 					 cpu_compress = 1;
 					 comp_bufs_count = rc;
@@ -1943,6 +1965,8 @@
 				SKYFS_ERROR("start client submit gpu compress  ino %llu write buf , i= %d, offset %llu, count %llu, last rc %llu, sizes_ptr %p \n", 
 						tmp->ino,match,   tmp->bufs[match].start,  SKYFS_OBJECT_NODE_SIZE, rc , sizes_ptr);
 				gettimeofday(&t1, NULL);
+
+				//Gpu_task_index = lock_gpu_comp_task(Gpu_task_num);
 				rc = __skyfs_C2O_gpu_compress_bufs(comp_type, 
 						&tmp->bufs[match],
 						tmp->bufs[match].buf, 
@@ -1951,6 +1975,9 @@
 					       	&sizes_ptr,
 						&comp_bufs_ptrs
 						);
+
+				//unlock_gpu_comp_task(Gpu_task_index);
+
 				gettimeofday(&t2, NULL);
 				Gpu_comp_time += (t2.tv_sec*1000000 + t2.tv_usec);
 				Gpu_comp_time -= (t1.tv_sec*1000000 + t1.tv_usec);
@@ -2224,12 +2251,26 @@
 		// set dev_id for all 2 bufs, because other comp_buf_parameters is set to zero by calloc
 		for (int n = 0 ; n< 2; n++){
 			new_compbuf->bufs[n].dev_id = (new_compbuf->ino+n) % 2;
+			int dev_idx =  new_compbuf->bufs[n].dev_id;
+			char * d_temp_ptr = NULL;
+ 			select_cuda_device(new_compbuf->bufs[n].dev_id);
+
+			int lock_idx = lock_gpu_init_task(dev_idx);
 			// important !! select cuda device and create streams 
- 			select_cuda_device(new_compbuf->bufs[n].dev_id);
-			create_cuda_streams(new_compbuf->bufs[n].dev_streams, 2);
 			new_compbuf->bufs[n].d_size = get_nvcomp_zstd_temp_size();
+			/*
+			alloc_cuda_memory(d_temp_ptr,  new_compbuf->bufs[n].d_size);
+			if(d_temp_ptr != NULL){
+				SKYFS_ERROR_1("Init alloc d_temp %p \n", d_temp_ptr);
+				 new_compbuf->bufs[n].d_temp = d_temp_ptr;
+			}else{
+				SKYFS_ERROR_1("Init alloc d_temp failed !\n");
+			}*/
 
+			create_cuda_streams(new_compbuf->bufs[n].dev_streams, 2);
+			unlock_gpu_init_task(lock_idx);
 
+
 		}
 		rc = register_host_data((void*)new_compbuf, sizeof(skyfs_C_gcompbuf_unit_t));
 		gettimeofday(&t2,NULL);
Index: client_compress_thread.c
===================================================================
--- client_compress_thread.c	(revision 267)
+++ client_compress_thread.c	(working copy)
@@ -11,6 +11,9 @@
 
 #define COMP_NUM_THREADS (64)
 #define COMP_NUM_TASKS (4)
+#define GCOMP_NUM_TASKS (66)
+//#define GCOMP_INDEX_RANGE (16)
+#define GCOMP_GROUP_RANGE (4)
 
 sem_t task_sem[COMP_NUM_THREADS];       // 任务同步信号量
 sem_t done_sem[COMP_NUM_THREADS];       // 任务完成信号量
@@ -18,8 +21,54 @@
 pthread_t compress_threads[COMP_NUM_THREADS];
 
 pthread_mutex_t compress_task_mutex_locks[COMP_NUM_TASKS];// 最大并发任务锁数组
+
+pthread_mutex_t gcompress_task_mutex_locks[GCOMP_NUM_TASKS];// 最大并发任务锁数组
 compress_task_t  comp_tasks[COMP_NUM_THREADS];
 
+
+pthread_mutex_t * get_dev_mutex(int dev_idx)
+{
+	int delta = random();
+	delta = delta % 5;
+	return  &gcompress_task_mutex_locks[dev_idx * 8 +16 + delta];
+}
+int lock_gpu_comp_task(int gtask_num)
+{
+	int start_index = (gtask_num/GCOMP_GROUP_RANGE) * GCOMP_GROUP_RANGE;
+	int this_index = start_index + (((int)random()) % GCOMP_GROUP_RANGE); 
+	pthread_mutex_lock( &gcompress_task_mutex_locks[this_index]);
+	return this_index;
+}
+
+void unlock_gpu_comp_task(int gtask_index)
+{
+
+	pthread_mutex_unlock( &gcompress_task_mutex_locks[gtask_index]);
+	return ;
+}
+
+int lock_gpu_init_task(int dev_idx)
+{
+	int delta = random();
+	int select_idx = 0;
+	delta = 0;
+	select_idx = dev_idx*8+32 +delta;
+	//int start_index = (gtask_num/GCOMP_GROUP_RANGE) * GCOMP_GROUP_RANGE;
+	//int this_index = start_index + (((int)random()) % GCOMP_GROUP_RANGE); 
+	pthread_mutex_lock( &gcompress_task_mutex_locks[select_idx]);
+	return select_idx;
+}
+
+void unlock_gpu_init_task(int gtask_index)
+{
+
+	pthread_mutex_unlock( &gcompress_task_mutex_locks[gtask_index]);
+	return ;
+}
+
+
+
+
 void* compress_thread_function(void* arg) {
    compress_task_t * this_task = NULL;
    int task_idx = 0;
@@ -95,6 +144,11 @@
 	for(int j = 0; j<COMP_NUM_TASKS;j++){
 		pthread_mutex_init(&compress_task_mutex_locks[j],NULL);
 	}
+
+	for(int j = 0; j<GCOMP_NUM_TASKS;j++){
+		pthread_mutex_init(&gcompress_task_mutex_locks[j],NULL);
+	}
+
 	
 	// 初始化压缩任务参数
 	
@@ -105,6 +159,8 @@
         	pthread_create(&compress_threads[n], NULL, compress_thread_function, (void*)(&comp_tasks[n]));
 
 	}
+
+	
 }
 
 int  run_multithead_compress(compress_task_t * comp_task)
@@ -121,7 +177,13 @@
 	int x = 0;
 	int rc = 0;
 	int complete_tasks = 0;
+	
 
+	if(tasks_per_thread == 0){
+		tasks_per_thread =1;
+	}
+
+
 	compress_task_t * curr_task = NULL;
 
 	// 设置所有参数到各个压缩线程
Index: client_ito.c
===================================================================
--- client_ito.c	(revision 267)
+++ client_ito.c	(working copy)
@@ -59,6 +59,8 @@
 
 extern int isComp_zstd();
 
+extern pthread_mutex_t * get_dev_mutex(int dev_idx);
+
  extern void free_cpu_batch(void * cpu_batch);
 #if 0
  extern void call_gpu_zstd_compress(char* data, int size,
@@ -1281,6 +1283,10 @@
 	void * pbatch_obj = NULL;
 	struct timeval tv1, tv2;
 	uint64_t durtime;
+	void * old_d_temp = gpubuf->d_temp;
+	size_t old_d_size = gpubuf->d_size;
+	int dev_idx = gpubuf->dev_id;
+	pthread_mutex_t * dev_mutex = get_dev_mutex(dev_idx);  
 
 	if (compress_type == 0x100){ // GPU_zstd_COMPRESS
 
@@ -1295,7 +1301,7 @@
 				&gpubuf->d_temp, &gpubuf->d_size,
 				gpubuf->dev_streams,
 				&gpubuf->uncomp_dev_buf, &gpubuf->uncomp_dev_ptrs, &gpubuf->uncomp_dev_sizes,
-				&gpubuf->comp_dev_buf, &gpubuf->comp_dev_ptrs, &gpubuf->comp_dev_sizes
+				&gpubuf->comp_dev_buf, &gpubuf->comp_dev_ptrs, &gpubuf->comp_dev_sizes, dev_mutex
 				);
 
 		gettimeofday(&tv2, NULL);
@@ -1358,9 +1364,11 @@
 			write_obj_do_compress_time += do_time;
 			write_obj_post_compress_time += post_time;
 
-			
+			if(old_d_temp == NULL)
+				SKYFS_ERROR_1("Gpu_nvcomp  ,pre_time %lu, d_tomp %p, d_size %lu, old_d_temp %p , old_d_size %lu\n ",
+					pre_time, gpubuf->d_temp, gpubuf->d_size, old_d_temp, old_d_size);
 
-			SKYFS_ERROR("gpu nvcomp compress data_ptrs %p , [0] %p \n ", data_ptrs, real_data_ptrs[0]);
+			//SKYFS_ERROR("gpu nvcomp compress data_ptrs %p , [0] %p \n ", data_ptrs, real_data_ptrs[0]);
 		        	
 
 		}	
